quickcheck-dynamic/src/Test/QuickCheck/StateModel.hs

```diff
@@ -123,63 +127,69 @@ class
   precondition :: state -> Action state a -> Bool
   precondition _ _ = True
 
-  -- | Postcondition on the `a` value produced at some step.
-  -- The result is `assert`ed and will make the property fail should it be `False`. This is useful
-  -- to check the implementation produces expected values.
-  postcondition :: state -> Action state a -> LookUp -> a -> Bool
-  postcondition _ _ _ _ = True
-
-  -- | Allows the user to attach information to the `Property` at each step of the process.
-  -- This function is given the full transition that's been executed, including the start and ending
-  -- `state`, the `Action`, the current environment to `Lookup` and the value produced by `perform`
-  -- while executing this step.
-  monitoring :: (state, state) -> Action state a -> LookUp -> a -> Property -> Property
-  monitoring _ _ _ _ = id
-
--- | Perform an `Action` in some `state` in the `Monad` `m`.  This
--- is the function that's used to exercise the actual stateful
--- implementation, usually through various side-effects as permitted
--- by `m`. It produces a value of type `a`, eg. some observable
--- output from the `Action` that should later be kept in the
--- environment through a `Var a` also passed to the `nextState`
--- function.
---
--- The `Lookup` parameter provides an /environment/ to lookup `Var
--- a` instances from previous steps.
-newtype RunModel state m = RunModel { perform :: forall a. state -> Action state a -> LookUp -> m a }
-
-type LookUp = forall a. Typeable a => Var a -> a
-
-type Env = [EnvEntry]
-
-data EnvEntry where
-  (:==) :: (Show a, Typeable a) => Var a -> a -> EnvEntry
+-- TODO: maybe it makes sense to write
+-- out a long list of these instances
+type family Realized (m :: Type -> Type) a :: Type
```

Comment:
Question: What's the rationale behind introducing this type family?

Reply:
C.f. discussion below about wanting different backends with different concrete representations of the things referenced in the model. Important if you want to use a model state that refers to things that would be indexed by the `s` in the state model type otherwise.

See the Registry example for a showcase of why this matters.

Reply:
Note that an example of this is `action` in the `DL` monad - it requires the result of a state model action to be `Typeable` - but if you want the result to be a `ThreadId (IOSim s)` you can't do it because `s` can't be typeable if you want to actually run your property.